### ISSUE TITLE
Picking up where left off, after a crash.  (Invalid Global Session)

### ISSUE THREAD

**Issue by fullshutter** [ndbroadbent/icloud_photos_downloader#44]
Hi @ndbroadbent ,

Sorry to be a frequent flyer!  Noticed something after I got an "Invalid Global Session" a couple of times.  I assume the Invalid Global Session is the result of a timeout from trying to download a large file (i.e. a large video)

When I re-execute, and I see the "already exists", jump 100 and check that file (forgive the pseudocode).  If the file that is "100 later" already exists, another 100 is added to the counter and the 201st file is checked.

Makes complete sense.  Only question I have is, what if the "n+100" file does indeed exist (in my case, because i downloaded all photos first, and am now running for videos), but the n+41 or n+73 or n+98 file does not?  Will I end up missing the ones that are missing, but are under the "100 mark"?

Let me know if that doesn't make sense and I'll try to summarize better.

Thanks!!

**Comment by ndbroadbent** [comment]
Hi @fullshutter, no problem! I think what is happening is that the `pyicloud` library requests "pages" of photos from the iCloud API, and each page contains 100 photos. We're also using the [tqdm](https://github.com/tqdm/tqdm) library to show the progress bar, and the progress is only updated every 10ms or so, so the delay that you're seeing is when we make the API requests to Apple's servers. But we do check that every single file exists before making the next API request.

Hope that makes sense!

**Comment by fullshutter** [comment]
Hi @ndbroadbent .  Awesome, makes sense.  Thank you.  Out of curiosity, if things timeout in the middle of a download (or if I issue the break character -- which I've done a couple of times when it has hung for at the same progress for several hours, with no network traffic).....when I re-run and it checks for the presence of a file, before it downloads it, is there a possibility that I will end up "short" on the full file that got timed out upon (or that I gave the "break" on)?

In other words, are there any file size checks on the files when seeing if they exist locally, in comparison with iCloud?  Certainly not expecting a hash or checksum.  That would be too heavy, I'm sure.  But curious if a crude/simple file sizes check is done, or if only file name 'existence'.

Thanks again for writing this.  Major props.  And for the quick replies!

**Comment by ndbroadbent** [comment]
That's a good point, we don't currently detect files that were only partially downloaded. I'm not sure if the iCloud API provides filesize information, but we could issue HEAD requests to get the content-length. Another way could be to just check that files are valid images on your local machine. There should be a third-party tool script or application that could scan your download photo and remove any corrupted images and videos.

Actually the best way is to probably download files to a temp directory, and issue a move command when we're sure that the file was downloaded successfully. That's not a bad idea.

**Comment by fullshutter** [comment]
Thanks @ndbroadbent .  Happy to contribute some ideas for next release. :-)

In the meantime, any recommendations on something open source/free for scanning the photos and video for corrupt items? 

     - Were you thinking something offline to do a basic check on them?
     - Or something "smart" that would actually check them against the source, in icloud?

**Comment by chadj** [comment]
pyicloud does provide file size information

```python
photo.size # original file size
photo.versions['medium']['size']
```

**Comment by ndbroadbent** [comment]
Thanks @chadj, I’ll use that to redownload any partial files.

**Comment by ndbroadbent** [comment]
Closing for now, PRs welcome if anyone wants to help with the "partial file redownloading" feature!
